[CLD-6324] Cloud IP Filtering

nickmisasi opened the pull request:
#### Summary
This PR adds support for setting IP Filters in Cloud workspaces. These IP Filters consist of a description, and an IP address range in CIDR notation. Setting these IP Filters means that only those with an IP address within the set range will be able to access the Mattermost workspace. 

Accompanying Enterprise PR can be found here: https://github.com/mattermost/enterprise/pull/1574

#### Ticket Link
https://mattermost.atlassian.net/browse/CLD-6319
https://mattermost.atlassian.net/browse/CLD-6324\

#### Screenshots
Filtering disabled:
![image](https://github.com/mattermost/mattermost/assets/12176405/e92b77b0-00b1-4f14-8d28-55c27d757c5e)

Filtering enabled, no rules:
![image](https://github.com/mattermost/mattermost/assets/12176405/822e7464-ff1c-4690-9afc-70d49bcc6390)

Add filter modal:
(with validation errors)
![image](https://github.com/mattermost/mattermost/assets/12176405/3d336add-cb7b-46b2-96da-d79a8c51cb06)
(without validation errors)
![image](https://github.com/mattermost/mattermost/assets/12176405/cfe1ba5f-b7d3-447a-bd92-367a94f1b4fa)

Edit filter modal:
(same modal as add modal just different title - validation errors are the same too)
![image](https://github.com/mattermost/mattermost/assets/12176405/c3bf4e25-923d-4d37-b461-a682b38dc7d3)

Filtering enabled, rules present:
![image](https://github.com/mattermost/mattermost/assets/12176405/99d3026c-99fb-41f7-8b73-f99f3f187a7e)
Edit/Delete buttons on hover of filter rows in table:
![image](https://github.com/mattermost/mattermost/assets/12176405/671406ee-d8d0-42e6-b127-bf1aa46433f2)


Added, edited, or deleted filters - toggle is enabled, and some rules still exist:
![image](https://github.com/mattermost/mattermost/assets/12176405/495a32cc-db5a-4407-a6c6-e4fa355a327d)

Rules existed, but all have been deleted (ie, none remaining, but toggle is ON), and the save button is clicked:
![image](https://github.com/mattermost/mattermost/assets/12176405/8b6ceef1-5ca9-4bb6-a5a0-67800c71d90b)

Enable IP Filtering toggle is turned OFF (regardless of existing rules or not):
![image](https://github.com/mattermost/mattermost/assets/12176405/1e42ba1b-9301-4617-ac33-d5e7b894280f)

IP Filtering rules set, but client's IP Address is not included in the rules:
![image](https://github.com/mattermost/mattermost/assets/12176405/a8a6874c-1eae-48f7-8c37-c5eccf1664a1)


#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates).
* API additions—new endpoint, new response fields, or newly accepted request parameters.
* Database changes (any).
* Schema migration changes. Use the [Schema Migration Template](https://docs.google.com/document/d/18lD7N32oyMtYjFrJKwsNv8yn6Fe5QtF-eMm8nn0O8tk/edit?usp=sharing) as a starting point to capture these details as release notes. 
* Websocket additions or changes.
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating).
* New features and improvements, including behavioral changes, UI changes, and CLI changes.
* Bug fixes and fixes of previous known issues.
* Deprecation warnings, breaking changes, or compatibility notes.

If no release notes are required, write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```
NONE
```
-->
```release-note
Support for IP Filtering in Cloud
```

nickmisasi commented:
@matthewbirtch You should be able to see the IP Filtering work via the spinwick link above. Some things I want to note:

1. The Add a filter button isn't as designed - I'll fix that in an upcoming commit
current: 
![image](https://github.com/mattermost/mattermost/assets/12176405/41b93098-808a-4511-958b-70dbfbbfc6fa)
designed:
![image](https://github.com/mattermost/mattermost/assets/12176405/15222104-5cf5-414d-b2d0-48c8b43762ab)
2. The switch selector for enabling/disabling IP Filtering is slightly different than that of the designs, as I re-used a component that was super uncooperative to me changing it without affecting other occurrences
3. Anywhere that links to documentation (Enable IP Filtering panel, add/edit filter modal, etc) doesn't actually link anywhere yet. Will need to figure out what pages to link to and adjust in a future commit
4. A note on the behaviour of enabling/disabling: Due to the nature of the backend not having an all-in-one enable/disable, the act of disabling simply sets all existing rules to disabled. This means that the page will treat no existing rules as if the feature itself is disabled. The page will change to the designed "NO IP Filtering rules added" pane, with the globe when you toggle the switch from OFF to ON (while no rules present), but leaving and coming back (without adding any rules and clicking save) will result in the page displaying OFF again.
5. In the confirmation modal after clicking save, it does not yet display the system administrator/owners email address, replaced with `your workspace owner`. I need to talk to security and look into the pheasability of  having that email address there (that email address can be changed and would need to be fetched, etc)
6. Work outstanding:
- The email sent to administrators when the IP Filters change. This one's a bit tricky, as we might need to throttle the number of emails sent (ie if I change them 15 times in 2 minutes while setting them up, we probably don't want 15 emails)
- The lockout screen has not been implemented in the CWS

Aside from the notes above - I think this should be ready for a first pass review from you.

I'll work on seeing if I can somehow get you access to the equivalent portal account to test out the "escape hatch" as well.

matthewbirtch commented on webapp/channels/src/components/admin_console/ip_filtering/ip_filtering.scss:
```diff
@@ -0,0 +1,406 @@
+                        padding: 16px;
+                        border: 1px solid var(--error-text-16, rgba(210, 75, 78, 0.16));
+                        background: var(--error-text-8, rgba(210, 75, 78, 0.08));
+                        background-color: var(--error-text-16, rgba(210, 75, 78, 0.16));
```
you should be able to remove this `background-color` line

matthewbirtch reviewed (changes requested):
This is looking great Nick! I went through a first pass of what I could. Just a few points of feedback from me beyond what you already called out:

1. In the empty state with the globe, the 'add a filter' text looks like it's sitting too low for some reason.
    <img width="334" alt="Screenshot 2023-10-10 at 6 31 49 PM" src="https://github.com/mattermost/mattermost/assets/2040554/300f6bcc-f3f3-4c88-9755-c5b86a1dfefa">


2. In the Add IP Filter modal, within the section notice "Your current IP address is...", can we use compass icons `icon-information-outline`. It's similar to what you have here, but is more aligned the rest of the icons in our icon system
    <img width="282" alt="image" src="https://github.com/mattermost/mattermost/assets/2040554/4a4a3cff-aee9-47d2-a531-5444313335fc">
    <img width="674" alt="image" src="https://github.com/mattermost/mattermost/assets/2040554/c648462e-f40d-4aae-b368-e6b87eb4dcab">

3. For the same notice, we are showing my current IP address is iPv6 format. Is that expected? I kind of assumed it would be IPv4, which I imagine is more commonly used? The example range in the help text that shows under the IP range field in this modal also implies IPv4. In addition, I was only able to successfully add a range in IPv4 format.

4. Similarly, my IPv4 is `99.255.35.42` and so I entered `99.255.35.42/24` as my range, but the system thinks my IP is not in range with the address range. And it's showing my IPv6 address in the warning notice.
    <img width="952" alt="image" src="https://github.com/mattermost/mattermost/assets/2040554/405fa9e3-2c28-4952-8c53-f455d98721cd">

5. The warning section notice "Your address is not included..." has its background color set to use 0.16 opacity, but it should be 0.08 opacity. I added a comment directly in the CSS for this.

6. Looks like the close button on the modal is positioned a bit off and the x icon is smaller than it's supposed to be. A comparison with the Create new channel modal for reference:
    <img width="636" alt="image" src="https://github.com/mattermost/mattermost/assets/2040554/596fd88e-a399-4e4e-9112-7d85dcfe1a58">
    <img width="636" alt="image" src="https://github.com/mattermost/mattermost/assets/2040554/2c422e53-59b1-454e-879b-79b5178c9337">

7. For the 'invalid' error message, can we update the message to "Invalid CIDR address range"

Because I couldn't get past the IP address warning above, I was not able to save and test out some of the other items (like confirmation modals). So I'll leave it at this for now. I can come back to test when I hear back from you about the IPv4/v6 conversation.

I'm not sure if we should consider some sort of toggle to use IPv4 or IPv6 or if we would be ok just supporting IPv4. I don't recall that coming up in the requirements.

nickmisasi commented:
/update-branch

nickmisasi commented:
/update-branch

mattermost-build commented:
Error trying to update the PR.
Please do it manually.

matthewbirtch commented on webapp/channels/src/components/admin_console/ip_filtering/ip_filtering.scss:
```diff
@@ -0,0 +1,406 @@
+            }
+
+            .AllowedIPAddressesSection {
+                display: inline-flex;
```
you shouldn't need any of these styles for this section:

~~overflow: hidden;~~
~~width: 920px;~~
~~background: white;~~
~~border-top-left-radius: 4px;~~
~~border-top-right-radius: 4px;~~

matthewbirtch commented on webapp/channels/src/components/admin_console/ip_filtering/ip_filtering.scss:
```diff
@@ -0,0 +1,406 @@
+
+            .TableSectionContent {
+                display: inline-flex;
+                width: 920px;
```
width: 100%;
~~min-height: 150px;~~
flex-direction: column;
padding: 20px;
~~background: #white;~~
~~border-radius: 8px;~~

matthewbirtch commented on webapp/channels/src/components/admin_console/ip_filtering/edit_section.tsx:
```diff
@@ -0,0 +1,200 @@
+                                                className='Button'
+                                                type='button'
+                                            >
+                                                {formatMessage({id: 'admin.ip_filtering.add_filter', defaultMessage: 'add a filter'})}
```
Add a filter

(capitalize A)

matthewbirtch commented on webapp/channels/src/components/admin_console/ip_filtering/ip_filtering.scss:
```diff
@@ -0,0 +1,406 @@
+                                display: flex;
+                                align-items: center;
+                                justify-content: center;
+                                padding: 8px;
```
let's adjust this to 
```
padding: 6px;
```

matthewbirtch reviewed (changes requested):
I've noted some more CSS changes directly in the code, and found a few other behavioral things beyond what I found earlier:

1. In the 'Add IP filter' modal, I can add an IP filter with just a name and no IP range right 
now.
2. I can add 2 filters with exactly the same range. Not a big deal, but I wonder if we should validate that to prevent duplicates of the exact same range.
3. I can also add 2 filters with exactly the same name - they should probably be validated to be unique as well
4. No matter what I try I can't seem to get the `Save` button in the footer to be enabled so I can't test the confirmation modals.

I need to chat with you further about the IPv4/6 topic, but wanted to call these things out as well

nickmisasi commented:
@matthewbirtch I've addressed the inline CSS feedback you've given me in the latest diff. 

>In the 'Add IP filter' modal, I can add an IP filter with just a name and no IP range right
now.

Fixed in latest diff.

>No matter what I try I can't seem to get the Save button in the footer to be enabled so I can't test the confirmation modals.

I am 99.9% sure the latest diff should be working for you - I was able to get the button to enable while spoofing the IPV6 locally. There was a small regression from a commit yesterday that I've fixed. 


> can add 2 filters with exactly the same range. Not a big deal, but I wonder if we should validate that to prevent duplicates of the exact same range.
I can also add 2 filters with exactly the same name - they should probably be validated to be unique as well

I'll look into these - since they're not super problematic, this may be something we can defer the fix of for a while

matthewbirtch commented on webapp/channels/src/components/admin_console/ip_filtering/ip_filtering.scss:
```diff
@@ -0,0 +1,402 @@
+            justify-content: flex-start;
+            padding: var(--spacing-xxl, 24px) 32px;
+            background-color: var(--center-channel-bg, #fff);
+            border-radius: 8px;
```
Small adjustments here and add the border+shadow
```
background-color: var(--center-channel-bg);
border-radius: 4px;
border: 1px solid rgba(var(--center-channel-color-rgb), 0.08);
box-shadow: 0px 2px 3px 0px rgba(0, 0, 0, 0.08);
```

matthewbirtch commented on webapp/channels/src/components/admin_console/ip_filtering/ip_filtering.scss:
```diff
@@ -0,0 +1,402 @@
+            flex-direction: column;
+            align-items: flex-start;
+            justify-content: flex-start;
+            padding: var(--spacing-xxl, 24px) 32px;
```
do we have these `--spacing` variables in place now? If not, we should remove

```
padding: 24px 32px;
```

matthewbirtch commented on webapp/channels/src/components/admin_console/ip_filtering/ip_filtering.scss:
```diff
@@ -0,0 +1,402 @@
+            padding: var(--spacing-xxl, 24px) 32px;
+            background-color: var(--center-channel-bg, #fff);
+            border-radius: 8px;
+            gap: var(--spacing-xxl, 24px);
```
Same here, remove the `--spacing` variable
```
gap: 24px;
```

matthewbirtch commented on webapp/channels/src/components/admin_console/ip_filtering/ip_filtering.scss:
```diff
@@ -0,0 +1,402 @@
+                flex-direction: column;
+                padding: 20px;
+                background: white;
+                border-radius: 8px;
```
remove 

~~background: white;~~
~~border-radius: 8px;~~

matthewbirtch reviewed (changes requested):
Confirming that these are fixed:

> > In the 'Add IP filter' modal, I can add an IP filter with just a name and no IP range right
> > now.
> 
> Fixed in latest diff.
> 
> > No matter what I try I can't seem to get the Save button in the footer to be enabled so I can't test the confirmation modals.
> 
> I am 99.9% sure the latest diff should be working for you - I was able to get the button to enable while spoofing the IPV6 locally. There was a small regression from a commit yesterday that I've fixed.

I also commented on a few more CSS items, plus another item I just noticed:

- If I click the edit button on a filter row, the modal title still says 'Add IP filter'. Can we update the title to 'Edit IP Filter'?

nickmisasi commented:
>If I click the edit button on a filter row, the modal title still says 'Add IP filter'. Can we update the title to 'Edit IP Filter'?

Fixed

>In the empty state with the globe, the 'add a filter' text looks like it's sitting too low for some reason.

Fixed

>In the Add IP Filter modal, within the section notice "Your current IP address is...", can we use compass icons icon-information-outline. It's similar to what you have here, but is more aligned the rest of the icons in our icon system

Fixed

>Looks like the close button on the modal is positioned a bit off and the x icon is smaller than it's supposed to be. A comparison with the Create new channel modal for reference:

Fixed. Also adjusted other modals that have the close button (delete confirmation, save confirmation, etc) 

>For the 'invalid' error message, can we update the message to "Invalid CIDR address range"

Fixed

### Confirmation Modals

I've made a few changes thanks to your callouts - mostly around a few bugs in which translation strings were used where. 

>Seeing this warning below after disabling IP filters, but this should only be shown for when I've adjusted IP filters (while still enabling them)

There was a bug for when you specifically have no rules, and then disable the toggle, the modal would not be incorrect. It will now show:
![image](https://github.com/mattermost/mattermost/assets/12176405/3a7619e5-55d7-4d4d-b555-260b9f9143a8)

>When disabling filters then saving, the correct confirmation modal should be:
Apply IP Filter Changes
Are you sure you want to apply these IP filter changes? There are currently no filters added, so all IP addresses will have access to the workspace.

The image you sent for this: 
![image](https://github.com/mattermost/mattermost/assets/12176405/f9d48dca-6098-4ee6-a95c-2fc081c3d354)

From my understanding is only shown when you delete all IP rules from the list, but don't disable the global toggle. 

When you disable the IP Filters at the global level (ie, with the toggle), you will see this error screen:

![image](https://github.com/mattermost/mattermost/assets/12176405/f295b889-66fc-48bd-8008-dffe29a40722)

>For the case where I've actually made changes to filters and kept it enabled, the confirmation modal should be more like the above, but we need to update the modal title and remove the additional {strong}

Based on the above two, I believe that the modal you pictured is correct as per the code - ie, if you edit or add filters (or go from disabled to enabled, with a filter present) and click save, you should see this:

![image](https://github.com/mattermost/mattermost/assets/12176405/e6841f7e-11a5-4c77-8416-f811df8eaf98)

#### In Summary

Here's a summary - let me know if this sounds right to you.

Add, edit, or delete IP Filters, but still have at least 1 rule, with the global toggle enabled, click save:
![image](https://github.com/mattermost/mattermost/assets/12176405/75a262f1-4dee-4db6-a8d4-c0dd45a8287f)

Have IP Filters, turn the global toggle to "OFF", click save:
![image](https://github.com/mattermost/mattermost/assets/12176405/0d1ab01f-d680-4ba5-bae4-cc5e5bf67dc8)

Delete all existing filters, leaving no filters, but leaving the global toggle to ON:
![image](https://github.com/mattermost/mattermost/assets/12176405/74fa03ec-69a4-4776-826d-fc33a7414200)

Delete all existing filters, leaving no filters, AND turning the global toggle to OFF:
![image](https://github.com/mattermost/mattermost/assets/12176405/44823dae-5982-4427-ba2e-49734abacb34)

I believe with my recent commit + this summary that all of your concerns around the confirmation modals have been addressed. Lmk!

nickmisasi commented:
@matthewbirtch I'd wait until a spinwick post about an updated server that comes after this comment before proceeding with review to be sure it's all up to date.

matthewbirtch commented on webapp/channels/src/components/admin_console/ip_filtering/save_confirmation_modal.scss:
```diff
@@ -0,0 +1,160 @@
+        font-size: 14px;
+        font-style: normal;
+        font-weight: 600;
+        line-height: 20px;
```
add
```
padding: 24px 32px;
```

nickmisasi commented:
/update-branch

matthewbirtch reviewed (approved):
Looks like the test server didn't update, but I'm going to approve to unblock this as it looks like the code changes were made. I can have a final look once the test server is up and running, but I'm good to approve this

nickmisasi commented:
/update-branch

nickmisasi commented:
@hmhealey @gabrieljackson this one should now be ready for review. I'll work on fixing the spinwick

hmhealey reviewed (changes requested):
I'm going to need some more time to review the rest of this since I only got about half way through, but here's some feedback on the stuff I've gone through so far

nickmisasi commented on server/channels/api4/api.go:
```diff
@@ -137,6 +137,8 @@ type Routes struct {
 
 	Drafts *mux.Router // 'api/v4/drafts'
+
+	IPFiltering *mux.Router // 'api/v4/ip_filtering'
```
The intention for this is that while the initial implementation is only Cloud, we could pretty easily expand for self hosted support down the road by more or less just implementing the interface - the UI would be re-usable

nickmisasi commented on server/public/model/cloud.go:
```diff
@@ -333,3 +333,13 @@ func (df *Feedback) ToMap() map[string]any {
 }
+
+type PatchInstallationRequest struct {
+	AllowedIPRanges  *AllowedIPRanges
```
Normally I would, but the Provisioner's practices don't actually use snake case anywhere, they simply write the struct field as it is - so leaving this means I don't need to have multiple structs with a "translation layer" before passing data onwards

hmhealey replied:
I guess we don't often have objects making their way from the web app through to the provisioner. I'd rather have the web app code match its conventions as well, so I feel like we're going to need a translation layer of some sort to make that work.

I don't know if anyone has any strong feelings about the casing of API parameters, but I'd guess the easiest way to do that would be to define an anonymous struct type for these in one of the API layers, right before the object is encoded. We could do something like
```go
var req *model.PatchInstallationRequest

type RequestJSON struct {
    AllowedIPRanges *model.AllowedIPRanges `json:"allowed_ip_ranges"`
}
    
json.NewEncoder(w).Encode((*RequestJSON)(req))
```

nickmisasi replied:
I'm not sure I follow - as is, it already works, because Go will just marshal the struct into JSON using the struct key (ie `AllowedIPRanges`). Having the types match that pascal casing makes everything line up. 

Or are you saying you'd prefer to have the webapp still conform to snake case?

hmhealey replied:
Yeah, I'd prefer to still send snake case to the web app and through the `api4` methods. I was never a huge fan of using snake case for it, but I'd rather stick to our standard behaviour than to end up with inconsistencies

nickmisasi commented on webapp/channels/src/components/admin_console/ip_filtering/edit_section.tsx:
```diff
@@ -0,0 +1,197 @@
+    }
+
+    return (
+        <div className='EditSection'>
```
I've split the edit section out into many components and addressed excessive nesting of divs as best I can

nickmisasi commented on webapp/channels/src/components/admin_console/ip_filtering/edit_section.tsx:
```diff
@@ -0,0 +1,197 @@
+
+    return (
+        <div className='EditSection'>
+            <div className='AllowedIPAddressesSection'>
+                <div className='SectionHeaderContent'>
+                    <div className='Frame1281'>
+                        <div className='TitleSubtitle'>
```
Copy from Figma I forgot to update

hmhealey replied:
We might want to be careful about leaning too much on that since I think there's likely cases where there's more semantic markup that we should be using to make sure stuff is accessible and behaves correctly

nickmisasi replied:
For sure. It's great for getting the basic layout on the page so you can work on back-of-front-end but needs optimizations afterwards

nickmisasi commented on webapp/channels/src/components/admin_console/ip_filtering/edit_section.tsx:
```diff
@@ -0,0 +1,197 @@
+                            {formatMessage(
+                                {
+                                    id: 'admin.ip_filtering.any_ip_can_access_add_filter',
+                                    defaultMessage: 'Any IP can access your workspace. To limit access to selected IP Addresses, {add}.',
```
Updated occurrences throughout to use rich formatting instead with FormattedMessage

nickmisasi commented on webapp/channels/src/components/admin_console/ip_filtering/add_edit_ip_filter_modal.scss:
```diff
@@ -0,0 +1,192 @@
+.IPFilteringAddOrEditModal {
+    .modal-dialog {
```
yeah without these the positioning is off and the style isn't quite right

nickmisasi commented:
/update-branch

hmhealey commented on webapp/channels/src/components/admin_console/ip_filtering/delete_confirmation.tsx:
```diff
@@ -0,0 +1,58 @@
+    filterToDelete?: AllowedIPRange;
+}
+
+export default function DeleteConfirmationModal({onClose, onConfirm, filterToDelete}: Props) {
```
I guess we'd ideally update all confirmation modals to match the same style, but I see your point. Maybe I'll make a follow up to make `ConfirmModal` match this style, and if you end up needing another confirmation modal elsewhere in the System Console, you can make this one more generic for that?

hmhealey commented on webapp/channels/src/components/admin_console/ip_filtering/add_edit_ip_filter_modal.tsx:
```diff
@@ -0,0 +1,156 @@
+    };
+
+    return (
+        <Modal
```
Please do use `GenericModal` going forward. We've been wanting for a while to get the modals updated to use something that isn't React Bootstrap and to use a more consistent design, and every time we add a new modal that uses React Bootstrap directly and has its own styling, that's another one we'll have to migrate manually

hmhealey commented on webapp/channels/src/components/admin_console/ip_filtering/add_edit_ip_filter_modal.tsx:
```diff
@@ -0,0 +1,156 @@
+
+import cidrRegex from 'cidr-regex';
+import React, {useState} from 'react';
+import {Button, Modal} from 'react-bootstrap';
```
Thanks!

nickmisasi commented:
/update-branch

hmhealey commented on webapp/channels/src/components/admin_console/ip_filtering/index.tsx:
```diff
@@ -0,0 +1,265 @@
+            setCurrentIPIsInRange(true);
+            return;
+        }
+        setCurrentIPIsInRange(ipFilters !== null && currentUsersIP !== null && isIPAddressInRanges(currentUsersIP, ipFilters));
```
Do we need to store `currentIPIsInRange` in a `useState` or can we just recalculate it every time this component is updated? Generally, we don't want to store derived values in state if we can avoid it

nickmisasi replied:
I've made this a derived value

hmhealey commented on webapp/channels/src/components/admin_console/ip_filtering/index.tsx:
```diff
@@ -0,0 +1,265 @@
+        if (!ipFilters?.length) {
+            return;
+        }
+        setFilterToggle(ipFilters?.some((filter: AllowedIPRange) => filter.Enabled === true) ?? false);
```
Same with `filterToggle` and `ipFilters`. Since they sort of depend on each other, it'll also avoid any potential desyncs or loops if we update `ipFilters` directly (instead of using `useEffect`s) when a filter is enabled/disabled or you click on the toggle button. That'll also prevent extra renders caused by one state change causing another state change

nickmisasi replied:
I need to know the previous value of the filterToggle between renders so that if the number of filters drops from 1 to 0 I can still have the toggle enabled and show the zero state. That's basically what the if statement in this is doing - if there are no filters, allow the toggle to independently set its value. If there are filters, then the value is based on if they are disabled or not.

Part of this is a quirk of the UX vs backend. There's no global enable/disable in the backend, it's per row.

hmhealey replied:
That seems a bit weird to me, but I guess that sort of thing happens in frontend coding where the UX doesn't quite line up. That's fine by me then

hmhealey commented on webapp/channels/src/components/admin_console/ip_filtering/index.tsx:
```diff
@@ -0,0 +1,265 @@
+                />
+            }
+            {
+                showAddModal &&
```
Instead of rendering these modals directly, it might be easier to open them with the the `openModal` action like
```
dispatch(openModal({
    dialogType: IPFilteringAddOrEditModal,
    dialogProps: {
        currentIP: currentUsersIP,
        // onClose can be omitted and replaced with the onExited prop that the ModalController provides
        onSave: handleEditFilter,
        existingRange: editFilter,
    },
}));
```

Doing that means you won't need state to store the visibility of the modal or its props in this component's state

hmhealey commented on webapp/channels/src/components/admin_console/ip_filtering/ip_filtering_utils.ts:
```diff
@@ -0,0 +1,54 @@
+
+        // Calculate the subnet mask bits
+        const subnetMaskBits = isIPv4 ?
+            (1 << 32 - subnetMaskNumber) - 1 :
+            (1 << 128 - subnetMaskNumber) - 1;
```
Is it safe to do `1 << 128` here? It seems like this is going to overflow, especially because JS doesn't have proper ints and every number is technically floating point

nickmisasi replied:
I think you're right. I've wrapped these calculations in a try-catch, which should prevent overflow and return false if it happens.

hmhealey commented on webapp/channels/src/components/admin_console/ip_filtering/save_confirmation_modal.test.tsx:
```diff
@@ -0,0 +1,75 @@
+
+import SaveConfirmationModal from './save_confirmation_modal';
+
+jest.mock('components/external_link', () => {
```
Why mock the `ExternalLink` component? If it's something to do with missing context, you can use `renderWithFullContext` from `test/react_testing_utils` instead of using React Testing Library directly

nickmisasi replied:
This is just for a shallow render. The `ExternalLink` component is redux connected to grab things like server ID/user ID/etc but I don't think having that fully qualified component in here adds any value to the test and adds a bunch of bloat in mock state

hmhealey replied:
You might be able to get away without that if you use `renderWithFullContext` here instead of `renderWithIntl`, but since our discussion on Friday got me to rename `renderWithFullContext` and use it everywhere (https://github.com/mattermost/mattermost/pull/25217), you don't need to bother changing this now. I'll make sure to hold that PR until this one goes in too since it'll definitely conflict with this

hmhealey commented on webapp/channels/src/components/admin_console/ip_filtering/save_confirmation_modal.tsx:
```diff
@@ -0,0 +1,90 @@
+                            {/* TODO - replace "workspace owner" with owner's email address? */}
+                            <div className='Subtitle'>
+                                {
+                                    formatMessage(
```
This should also use a `FormattedMessage` to only need a single translation string

hmhealey commented on webapp/channels/src/components/widgets/icons/fa_next_icon.tsx:
```diff
@@ -7,12 +7,14 @@ import {defineMessage} from 'react-intl';
 
 const iconTitle = defineMessage({
-    id: 'generic_icons.next',
+    id: t('generic_icons.next'),
```
Is this needed? I think `mmjstool` can properly detect i18n strings using `defineMessage`

nickmisasi replied:
I thought so too but the pipeline wouldn't pass without it: https://github.com/mattermost/mattermost/actions/runs/6627549647/job/18002705272

I think it might have broken from a recent merge or something. Not sure why it wasn't happening to anyone else though

hmhealey replied:
That's odd. I guess that's another reason for us to get away from our custom i18n tooling and `mmjstool` then towards the standard FormatJS stuff

nickmisasi commented on webapp/channels/src/components/admin_console/ip_filtering/edit_section/edit_section_no_filters_panel.tsx:
```diff
@@ -0,0 +1,46 @@
+
+const NoFiltersPanel = ({setShowAddModal}: NoFiltersPanelProps) => (
+    <div className='NoFilters'>
+        <div className='icon'>
```
I added the class but it didn't need styling so never added any. Removed the className from the div.

hmhealey commented on webapp/channels/src/components/admin_console/ip_filtering/index.tsx:
```diff
@@ -0,0 +1,275 @@
+            dialogType: IPFilteringAddOrEditModal,
+            dialogProps: {
+                currentIP: currentUsersIP!,
+                onClose: () => dispatch(closeModal(ModalIdentifiers.IP_FILTERING_ADD_EDIT_MODAL)),
```
The `ModalController` automatically passes an `onExited` prop to each modal that it mounts, which dispatches `closeModal` with the `modalId` passed from here. If you use that, you don't actually don't need any of these `onClose` props

nickmisasi replied:
Nice! Done.

hmhealey commented on webapp/channels/src/actions/admin_actions.jsx:
```diff
@@ -91,6 +91,33 @@ export async function samlCertificateStatus(success, error) {
     }
 }
 
+export async function getIPFilters(success, error) {
```
Ideally, you'd have the components dispatch the actions you added to `mattermost-redux` directly instead of wrapping them in this old style of Flux actions with success/error callbacks which bind to the global instance of the Redux store this way, but since they're consistent with the rest of this file file, it's fine to leave as-is.

I'm honestly surprised this file has survived for so long as-is since this is how we made all the existing Flux actions work with Redux when we first made the change to it. I guess Apps/ICU/Web Platform has always been focused more on the non-System Console app when its come to cleanup 😅

nickmisasi replied:
lol it seems given 2 patterns to follow I'll happen to pick the wrong one... Every time :P 

i went this route because it seemed more correct than wrapping some anonymous function to allow for async/await in the useEffect. Thinking about it now, I guess I could have just had useEffect trigger an async function and have that perform the await. If you're ok with this way to match the rest of the file I'm cool to leave it

hmhealey replied:
Yep, that's fine by me

hmhealey reviewed (changes requested):
The above comments were also supposed to be part of this

gabrieljackson reviewed (approved):
Very nice!

esarafianou reviewed (approved):
Looked at the usage of the cidr npm library and the new permissions in the APIs. That was not a thorough review.

jupenur commented:
I'm going to withdraw my request for review because I don't have time to review this any time soon, but be warned: I can see the other reviews were rushed and this code is not production-grade. Merging it as it is *will* result in downtime. I also want to point out that although this PR has 5 approvals, the associated PRs got far less attention.

spirosoik commented:
> I'm going to withdraw my request for review because I don't have time to review this any time soon, but be warned: I can see the other reviews were rushed and this code is not production-grade. Merging it as it is _will_ result in downtime. I also want to point out that although this PR has 5 approvals, the associated PRs got far less attention.

Could you please elaborate a bit more here? The statement  `Merging it as it is _will_ result in downtime.` is kind of unclear.

Just as a radar, we already ran a Production Readiness Review [here](https://community.mattermost.com/private-core/channels/prr-enterprise-ip-filtering) with @iyampaul and there were 159 comments addressed on this PR and for the associated PRs there were four-eyes (2 approvals) as we guide in our developer docs too.

jupenur reviewed (dismissed):
Here's a quick sample of the issues in this code. I literally only looked at this part I commented on, so don't assume the rest of it is ok. The reason I say this will result in downtime is that it's blocking IP addresses and if we get that wrong, admins are going to block themselves. I know this part I commented on is only the client-side validation, but that's the part admins are going to be relying on before actually committing their changes.

nickmisasi commented:
@jupenur I appreciate your concern around this function and thanks for the context. I want to provide some context of my own here:

First, there is some obvious inherent risk with a feature like this - having users lock themselves out of their workspace/instance is not just a possibility, but likely a certainty. This is a risk that we have accepted and planned for. We've worked very closely with UX to develop a few different mechanisms to aid in reducing any potential downtime:
- [The customer portal escape hatch](https://github.com/mattermost/customer-web-server/pull/1234) - the portal is hosted outside of the IP Filtering rules applied from this page. We've added what we refer to as an "escape hatch" where if you do lock yourself out of your instance you can go to disable IP Filtering to regain access. This escape hatch is referenced in many places throughout the UX:

- [When setting IP Filters, users are notified of the riskiness of the move, as well as given a link to the Customer Portal page to disable IP Filtering if necessary](https://user-images.githubusercontent.com/12176405/275615315-495a32cc-db5a-4407-a6c6-e4fa355a327d.png) 
- [An email is sent to all system administrators every time the filters change, including a mention of the user completeing the action as well as a link to the customer portal page to reset the IP Filters if necessary](https://user-images.githubusercontent.com/12176405/279166851-f993f412-fe6e-484e-a557-b02d352e37c1.png)

Re: the isIpAddressInRanges function.

Your highlighting of potential overflows here are correct - I had played around with this for a few days last month (and a while today as well), and based on testing believed I'd gotten the function to a stable point. Some things I'll mention:
- The UX flow doesn't allow someone to add an invalid IPV4/IPV6 address, so inputs to this function should at least be proper CIDR blocks
- IP addresses that are added are further validated [before they're actually applied on the backend in the provisioner](https://github.com/mattermost/mattermost-cloud/blob/b0f93ff3c27f58b9fef02dae369952a10ee471f9/model/installation_request.go#L549)

I've adjusted the `isIpAddressInRanges` function to leverage a 3rd party library in response to your eventual feedback. This 3rd party library ipaddr.js is different from the `cidr-regex` library, but I was able to use it to also replace the `cidr-regex` library. I'm hoping either you or @esarafianou would be able to approve the use of this so we can proceed with the feature. 

One last thing to note: this feature is locked behind a feature flag, and I've already written up an [extensive testing plan](https://docs.google.com/document/d/1bM88G_sXk1MeKYR-rpX4NFP195bHJq1ttGvWLjDklz4/edit) for QA post-merge. This feature won't be enabled until we're satisfied with the stability and user experience - if things come up, we will iterate.

esarafianou commented:
@nickmisasi thank you for such a detailed response. Using the `ipaddr.js` is ok. 

Is there a plan to add tests for this functionality? I see there are tests in the corresponding enterprise PR but not in this one.

nickmisasi commented:
@esarafianou in the latest commit I've added a couple server tests. It's tough to have truly meaningful tests given the dependency on the CWS + provisioner but what I've added should at least cover permissions, licensing, feature flags, and happy paths

nickmisasi commented:
/update-branch

nickmisasi commented:
/update-branch
